File: bindings/ios/Sources/Models/ModelConfig.swift

/*
 typedef struct {
     int32_t n_ctx; // text context, 0 = from model
     int32_t n_threads; // number of threads to use for generation
     int32_t n_threads_batch; // number of threads to use for batch processing
     int32_t n_batch; // logical maximum batch size that can be submitted to llama_decode
     int32_t n_ubatch; // physical maximum batch size
     int32_t n_seq_max; // max number of sequences (i.e. distinct states for recurrent models)
     ml_Path chat_template_path; // path to chat template file, optional
     const char *chat_template_content; // content of chat template file, optional
 } ml_ModelConfig;

 */
import NexaBridge

public struct ModelConfig: Codable {
    public var nCtx: Int32
    public var nThreads: Int32
    public var nThreadsBatch: Int32
    public var nBatch: Int32
    public var nUbatch: Int32
    public var nSeqMax: Int32

    var chatTemplatePath: String?
    var chatTemplateContent: String?
    
    public static let `default`: ModelConfig = {
        let mlConfig = ml_model_config_default()
        return .init(
            nCtx: 2048,
            nThreads: mlConfig.n_threads,
            nThreadsBatch: mlConfig.n_threads_batch,
            nBatch: mlConfig.n_batch,
            nUbatch: mlConfig.n_ubatch,
            nSeqMax: mlConfig.n_seq_max,
            chatTemplatePath: nil,
            chatTemplateContent: nil
        )
    }()
}

File: bindings/android/app/src/main/java/com/geniex/sdk/bean/ModelConfig.kt
package com.geniex.sdk.bean

/**
 * Model configuration corresponding to the native `geniex_ModelConfig` struct.
 */
data class ModelConfig(
    /** Text context size, 0 = use model default */
    var nCtx: Int = 2048,

    /** Number of threads used for text generation */
    var nThreads: Int = 8,

    /** Number of threads used for batch processing */
    var nThreadsBatch: Int = 8,

    /** Maximum logical batch size submitted to llama_decode */
    var nBatch: Int = 2048,

    /** Maximum physical batch size supported by backend */
    var nUBatch: Int = 512,

    /** Maximum number of distinct sequences (states) */
    var nSeqMax: Int = 1,

    /**
     * Number of layers to offload to GPU / NPU; -1 (the default) means all
     * layers, which llama.cpp interprets natively. The JNI layer forces 0
     * when [InputPluginBase.compute_unit] is [ComputeUnitValue.CPU] (and for
     * qairt, which ignores it). For [ComputeUnitValue.GPU] / [ComputeUnitValue.NPU]
     * / [ComputeUnitValue.HYBRID] the caller's value passes through.
     */
    var nGpuLayers: Int = -1,

    /** Path to the chat template file (optional) */
    val chat_template_path: String = "",

    /** Content of the chat template file (optional) */
    val chat_template_content: String = "",

    /** Speculative decoding type(s), comma-separated (llama_cpp only; "" / "none" = disabled).
     *  Draft models: "draft-mtp","draft-eagle3","draft-simple". Self-speculative: "ngram-*". */
    val spec_type: String = "",

    /** Draft GGUF for draft-* spec types ("" for ngram-* or when disabled) */
    val spec_draft_model: String = "",

    /** Max draft tokens per step (0 = plugin default of 3) */
    val spec_n_max: Int = 0,

    /** Min draft tokens per step (0 = llama.cpp default) */
    val spec_n_min: Int = 0,

    /** Min greedy draft probability (0 = llama.cpp default) */
    val spec_p_min: Float = 0.0f,

    /** Unified HTP power/clock-management mode, shared by qairt and llama_cpp
     *  ("low_power_saver".."burst"); "" / "default" = burst. */
    val power_mode: String = "",
)
